### PR TITLE
Update Keybase to 1.0.18-20170208165022,dc66461

### PR DESCRIPTION
**Pull request by joshka**
Includes new Keybase chat feature.
See https://keybase.io/blog/keybase-chat for announcement details.

Add pre and post flight install / uninstall scripts.

- [x] `brew cask audit --download {{cask_file}}` is error-free.
- [x] `brew cask style --fix {{cask_file}}` reports no offenses.
- [x] The commit message includes the cask’s name and version.

### CODE REVIEW

File: Casks/keybase.rb [hunk 1]
```diff
@@ -1,14 +1,24 @@
...
   app 'Keybase.app'
+
+  postflight do
+    system_command '/Applications/Keybase.App/Contents/Resources/KeybaseInstaller.app/Contents/MacOS/Keybase',
+                   args: ['--app-path=/Applications/Keybase.app', '--run-mode=prod', '--timeout=10']
+  end
+
+  uninstall_preflight do
+    system 'killall', '-kill', 'Keybase'
+    system '/Applications/Keybase.app/Contents/SharedSupport/bin/keybase', 'uninstall'
```
Comment (miccal) [inline Casks/keybase.rb]: Please use `system_command` instead of `system` in the `uninstall_preflight`, as you did in the `postflight`.

**Comment by miccal** [comment]
@joshka you seem to updated an old version of `keybase.rb` which is causing a merge conflict.

File: Casks/keybase.rb (hunk 1 above)
Comment (joshka) [inline Casks/keybase.rb]: This is intentional. I do not want the killall / uninstall command to fail when the Keybase app is not running or has been removed.

**Comment by joshka** [comment]
@miccal fixed the merge conflict

File: Casks/keybase.rb (hunk 1 above)
Comment (miccal) [inline Casks/keybase.rb]: `system` has been deprecated; you must use `system_command`. See discussion [here](https://github.com/caskroom/homebrew-cask/issues/23070).

Comment (joshka) [inline Casks/keybase.rb]: I understand that's the policy. I'd like to be consistent with it, but it fails when Keybase is not running. There is a `:must_succeed` argument to `system_command`, but passing `false` has no effect.

If I remove the killall and Keybase is running during uninstall, a bar icon continues to be displayed. Quitting using Cmd+Q does not dismiss the application. Only quitting from the dock gets rid of the open application / bar icon.

TL;DR I need a way of running system_command that does not fail when the command fails. Until that's available, this is the best way to run these commands and prevent the orphaned icon / app.

Comment (reitermarkus) [inline Casks/keybase.rb]: You can check if it is running first, and only then kill it.

```ruby
unless system_command('launchctl', args: ['list']).stdout.lines.grep(/keybase.Electron/).empty?
  # kill it
end
```

Comment (vitorgalvao) [inline Casks/keybase.rb]: I never understood (and this is not a criticism but a serious question) why people “TL;DR” *at the end* of a post. If a post is so long I don’t want to read it, I’m not going to scroll to the very end to see if by any chance that person had a “TL;DR”, I just won’t read it.

Why not just put a “TL;WR” (too long; won’t read) or summary at the start?

Comment (joshka) [inline Casks/keybase.rb]: I often use TL;DR to signify approximately "thanks for reading the reasons above, pay attention I'm asking a question down here that gets to the heart of the matter now that you have the context"

TL;DR "pay attention to this bit" ;)

**Comment by joshka** [comment]
Review comments all addressed. I changed the regex necessary to only include running keybase.Electron instances not stopped instances.

File: Casks/keybase.rb (hunk 1 above)
Comment (vitorgalvao) [inline Casks/keybase.rb]: @joshka Ah, so you use it as a sort of “in conclusion”.

File: Casks/keybase.rb [hunk 2]
```diff
@@ -1,14 +1,27 @@
...
   app 'Keybase.app'
+
+  postflight do
+    system_command "#{appdir}/Keybase.App/Contents/Resources/KeybaseInstaller.app/Contents/MacOS/Keybase",
+                   args: ["--app-path=#{appdir}/Keybase.app", '--run-mode=prod', '--timeout=10']
+  end
+
+  uninstall_preflight do
+    if system_command('launchctl', args: ['list']).stdout =~ %r{/^\d+.*keybase.Electron/}
+      system_command 'killall', args: ['-kill', 'Keybase']
```
Comment (vitorgalvao) [inline Casks/keybase.rb]: Why is this in an `uninstall_preflight` instead of a regular `uninstall`? It can take care of quitting apps.

File: Casks/keybase.rb [hunk 3]
```diff
@@ -1,14 +1,27 @@
...
+    end
+    system_command "#{appdir}/Keybase.app/Contents/SharedSupport/bin/keybase",
```
Comment (vitorgalvao) [inline Casks/keybase.rb]: On that same token, why isn’t this script handled by an `uninstall` as well?

Comment (vitorgalvao) [inline Casks/keybase.rb]: Also, isn’t `Keybase.app/Contents/SharedSupport/bin/keybase` a useful CLI? Shouldn’t it be added in a `binary`?

Comment (joshka) [inline Casks/keybase.rb]: The app stanza runs before the uninstall stanza, so the binary gets removed and hence is not able to run in the uninstall stanza. Killing the app first is required before running the keybase uninstall command.

This app *is indeed* a useful CLI. It is installed to the right place (/usr/local/bin) during either the first run of Keybase.app, or by the command I added in postflight stanza above.

installer / first run installs services, cli, and taskbar app (maybe some more stuff). Requires app to be in /Applications or ~/Applications already hence cannot be run from a script stanza.
uninstall command requires that the app still exists in /Applications, hence cannot be run from an uninstall stanza.

**Comment by joshka** [comment]
Also, it seems like this should deprecate the homebrew formula of the same name as it installs both the App and the CLI. The fact that keybase.io/download no longer ships a CLI only version of Keybase supports this. Once this merge goes through, I'll request a PR to drop the keybase formula with a reference to the cask.

**Comment by gabriel** [comment]
You don't need to run the `KeybaseInstaller.app` postflight because it happens automatically on app start.

Also, the uninstall can call `/Applications/Keybase.app/Contents/SharedSupport/bin/keybase uninstall` which will properly remove all the things it installs.

**Comment by joshka** [comment]
@gabriel yes. You're right. However by waiting until first run of the app start you have to run the app to get the command line to work.
The uninstall preflight does call the command you mentioned. See line 24

**Comment by gabriel** [comment]
Oops, yes... 

Thanks for setting up the cask!

**Comment by joshka** [comment]
No problem :)

File: Casks/keybase.rb [hunk 4]
```diff
@@ -1,14 +1,27 @@
...
 
   app 'Keybase.app'
+
+  postflight do
+    system_command "#{appdir}/Keybase.App/Contents/Resources/KeybaseInstaller.app/Contents/MacOS/Keybase",
+                   args: ["--app-path=#{appdir}/Keybase.app", '--run-mode=prod', '--timeout=10']
+  end
+
+  uninstall_preflight do
+    if system_command('launchctl', args: ['list']).stdout =~ %r{/^\d+.*keybase.Electron/}
```
Comment (gabriel) [inline Casks/keybase.rb]: On Sierra, launchd list presents Keybase app as:

```
> launchctl list | grep keybase
2166	0	keybase.Electron.19224
```

which fails this regex.

Comment (joshka) [inline Casks/keybase.rb]: I'm not sure that it fails that regex. http://rubular.com/r/xXYsVBMOck

Comment (gabriel) [inline Casks/keybase.rb]: Ahh maybe failing another way:

- Install Keybase (not from cask, from https://keybase.io/docs/the_app/install_macos)
- Run Keybase.app
- Run `brew cask install keybase`

It doesn't kill the running Keybase.

Comment (joshka) [inline Casks/keybase.rb]: The uninstall preflight stanza only runs when uninstalling a version of keybase installed by cask. The regex works fine for cask installed keybase.

This behavior seems like it's more of a HBC bug rather than a bug of this cask. Perhaps HBC should stop the install rather than removing an existing app at that location / attempting to overwrite. I'll raise an issue to talk over this.

Here's the output that I see when attempting what you've described.

```
$ brew cask install keybase
==> Downloading https://prerelease.keybase.io/darwin/Keybase-1.0.18-20170209165653%2B17b641d.dmg
Already downloaded: /Users/joshka/Library/Caches/Homebrew/Cask/keybase--1.0.18-20170209165653,17b641d.dmg
==> Verifying checksum for Cask keybase
==> Removing App: '/Applications/Keybase.app'
Error: It seems there is already an App at '/Applications/Keybase.app'.
Error: nothing to install
```

To get past the issue, right click the icon in the dock and quit keybase.

Comment (gabriel) [inline Casks/keybase.rb]: This is what I see:

```sh
kbmb ➜ ~ brew cask install keybase
==> Downloading https://prerelease.keybase.io/darwin/Keybase-1.0.18-20170209165653%2B17b641d.dmg
Already downloaded: /Users/gabe/Library/Caches/Homebrew/Cask/keybase--1.0.18-20170209165653,17b641d.dmg
==> Verifying checksum for Cask keybase
==> 2017-02-15T16:40:04.177053-08:00 ▶ [INFO keybase launchd.go:291] 001 Removing /Users/gabe/Library/LaunchAgents/keybase.kbfs.plist
==> 2017-02-15T16:40:04.181578-08:00 ▶ [INFO keybase launchd.go:217] 002 Waiting for keybase.kbfs to exit...
==> 2017-02-15T16:40:04.391075-08:00 ▶ [INFO keybase launchd.go:160] 003 Stopped keybase.kbfs
==> 2017-02-15T16:40:04.394875-08:00 ▶ [INFO keybase mounter_nix.go:29] 004 Trying to unmount: /keybase
==> 2017-02-15T16:40:04.407014-08:00 ▶ [INFO keybase launchd.go:291] 005 Removing /Users/gabe/Library/LaunchAgents/keybase.service.plist
==> 2017-02-15T16:40:04.412145-08:00 ▶ [INFO keybase launchd.go:217] 006 Waiting for keybase.service to exit...
==> 2017-02-15T16:40:04.624894-08:00 ▶ [INFO keybase launchd.go:160] 007 Stopped keybase.service
==> 2017-02-15T16:40:04.629004-08:00 ▶ [INFO keybase launchd.go:291] 008 Removing /Users/gabe/Library/LaunchAgents/keybase.updater.plist
==> 2017-02-15T16:40:04.632981-08:00 ▶ [INFO keybase launchd.go:217] 009 Waiting for keybase.updater to exit...
==> 2017-02-15T16:40:04.845010-08:00 ▶ [INFO keybase launchd.go:160] 00a Stopped keybase.updater
==> 2017-02-15T16:40:04.970214-08:00 ▶ [INFO keybase install_osx.go:694] 00b Removing KBFuse
==> 2017-02-15T16:40:05.055503-08:00 ▶ [INFO keybase install_osx.go:699] 00c Removing privileged helper tool
==> 2017-02-15T16:40:05.133359-08:00 ▶ [INFO keybase install.go:222] 00d Removing /usr/local/bin/keybase
Error: It seems there is already an App at '/Applications/Keybase.app'.
Error: nothing to install
```

Comment (joshka) [inline Casks/keybase.rb]: Let's continue this over in #30142
